src/build/dlopen.c

```diff
@@ -0,0 +1,130 @@
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <dlfcn.h>
+#include <gnumake.h>
+
+#ifdef __APPLE__
+#  define DL_EXTENSION ".dylib"
+#else
+#  include <linux/limits.h>
+#  define DL_EXTENSION ".so"
+#endif
+
+char		*mk_dlpath(char const *nm, unsigned int argc, char **argv);
+char		*mk_dlerror(char const *nm, unsigned int argc, char **argv);
+int		dlopen_gmk_setup(void);
+static char	*getpath(char *dl_pathname, void *handle);
+
+int plugin_is_GPL_compatible;
+
+/** GNU make-callable function to return the absolute pathname of a dynamic library.
+ * 
+ * @return NULL on failure; otherwise, a string. 
+ * 
+ * @note argv[0] should point to the name of the library minus the extension.
+ */
+char * mk_dlpath(UNUSED char const *nm, UNUSED unsigned int argc, char **argv)
+{
+	void	*handle;
+	char	*extended_file_name;
+	char	*dl_pathname;
+
+	if (argv[0] == NULL) {
+		fprintf(stderr, "file name NULL\n");
+		return NULL;
+	}
+	
+	extended_file_name = gmk_alloc(1 + strlen(DL_EXTENSION) + strlen(argv[0]));
+	sprintf(extended_file_name, "%s%s", argv[0], DL_EXTENSION);
+
+	handle = dlopen(extended_file_name, RTLD_LAZY | RTLD_LOCAL);
+	if (!handle) {
+		/* fprintf(stderr, "%s\n", dlerror()); */
+		gmk_free(extended_file_name);
+		return NULL;
+	}
+
+	dl_pathname = getpath(extended_file_name, handle);
+
+	gmk_free(extended_file_name);
+	dlclose(handle);
+	return dl_pathname;
+}
+
+
+#ifdef __APPLE__
+static char *getpath(char *extended_file_name, void *handle)
+{
+	// Iterate through all images currently in memory
+	for (int32_t i = _dyld_image_count(); i >= 0 ; i--) {
+		// dlopen() each image, check handle
+		char const *image_name = _dyld_get_image_name(i);
+		uv_lib_t *probe_lib = jl_load_dynamic_library(image_name, JL_RTLD_DEFAULT);
+		void *probe_handle = probe_lib->handle;
+		uv_dlclose(probe_lib);
+
+		// If the handle is the same as what was passed in (modulo mode bits), return this image name
+		if (((intptr_t)handle & (-4)) == ((intptr_t)probe_handle & (-4))) {
+			char *dl_pathname = gmk_alloc(strlen(image_name) + 1);
+			strcpy(dl_pathname, image_name);
+			return dl_pathname;
+		}
+	}
+	return NULL;
+}
+#else
+static char *getpath(char *extended_file_name, void *handle)
+{
+	char	*dl_pathname = gmk_alloc(PATH_MAX);
+	char	*base;
+
+	if (dlinfo(handle, RTLD_DI_ORIGIN, dl_pathname) < 0) {
+		/* fprintf(stderr, "%s\n", dlerror()); */
+		gmk_free(dl_pathname);
+		return NULL;
+	}
+
+	/*
+	 * dlinfo() just gives the absolute path down to and including the
+	 * directory the file is in, so we have to append the file name itself
+	 * preceded by the '/' separator.
+	 */
+	base = strrchr(extended_file_name, '/');
+	if (base) {
+		strcat(dl_pathname, base);
+	} else {
+		strcat(dl_pathname, "/");
+		strcat(dl_pathname, extended_file_name);
+	}
+	return dl_pathname;
+}
+#endif
+
+char *mk_dlerror(UNUSED char const *nm, UNUSED unsigned int argc, UNUSED char **argv)
+{
+	char *msg = dlerror();
```

Comment:
The issue may be that the string returned from `dlerror()` changes for every call to the `dl*` functions.  So if GNU make calls `dlopen()` itself after the call to `mk_dlopen()`, then the error string is trashed.

You may need to call `dlerror()` from the `mk_dlpath()` function, and then cache the string yourself.

Or verify that your `mk_dlerror()` function is actually being called from `$(dlerror )`.  Maybe via `fprintf(stderr, ...)`

GNU make _shouldn't_ generally be calling `dlopen()` itself, but who knows what it does. :(

Reply:
True. I've corrected the "what if dlopen() returns NULL" issue, and confirmed that I'm not fprintf-ing dlerror(), and strangely now

```
$(info $(dlpath $(BUILD_DIR)/lib/.libs/no-such-file))
$(info $(dlerror ))

gives me 

```
<blank line because I handed it a bogus value>
build/lib/.libs/no-such-file.so: cannot open shared object file: No such file or directory
```

which is what I thought would happen (but didn't when I didn't leave a space in $(dlerror)). Confirmed that it's really calling dlerror().